addons/mail/models/mail_thread.py

```diff
@@ -1581,26 +1624,77 @@ def message_parse(self, message, save_original=False):
                 stored_date = datetime.datetime.now()
             msg_dict['date'] = stored_date.strftime(tools.DEFAULT_SERVER_DATETIME_FORMAT)
 
-        parent_ids = False
-        if msg_dict['in_reply_to']:
-            parent_ids = self.env['mail.message'].search(
-                [('message_id', '=', msg_dict['in_reply_to'])],
-                order='create_date DESC, id DESC',
-                limit=1)
-        if msg_dict['references'] and not parent_ids:
-            references_msg_id_list = tools.mail_header_msgid_re.findall(msg_dict['references'])
-            parent_ids = self.env['mail.message'].search(
-                [('message_id', 'in', [x.strip() for x in references_msg_id_list])],
-                order='create_date DESC, id DESC',
-                limit=1)
-        if parent_ids:
-            msg_dict['parent_id'] = parent_ids.id
-            msg_dict['is_internal'] = parent_ids.subtype_id and parent_ids.subtype_id.internal or False
-
-        msg_dict.update(self._message_parse_extract_payload(message, save_original=save_original))
+        parent_message = self._get_parent_message(msg_dict)
+        if parent_message:
+            msg_dict['parent_id'] = parent_message.id
+            msg_dict['is_internal'] = parent_message.subtype_id and parent_message.subtype_id.internal
+
+        msg_dict['is_bounce'] = self._message_parse_is_bounce(message, msg_dict)
+        msg_dict.update(self._message_parse_extract_payload(message, msg_dict, save_original=save_original))
         msg_dict.update(self._message_parse_extract_bounce(message, msg_dict))
```

Comment:
Should we only do this if "is_bounce" then? Or do we want to extract bounce values even if bounce detection says it's not a bounce.

Might change the flow compared to before and have less keys though, not sure.

If we need to do it all the time, then a small explanation would not hurt.